### ISSUE TITLE
Ensure Index Behaviour 

### ISSUE THREAD

**Issue by netag** [Litote/kmongo#207]
Hi all

I'm having some trouble understanding the reasoning behind the [implementation](https://github.com/Litote/kmongo/blob/master/kmongo-core/src/main/kotlin/org/litote/kmongo/MongoCollections.kt#L634-L647) of `ensureIndex`. 

```kotlin
fun <T> MongoCollection<T>.ensureIndex(keys: Bson, indexOptions: IndexOptions = IndexOptions()): String {
    return try {
        createIndex(keys, indexOptions)
    } catch (e: MongoCommandException) {
        //there is an exception if the parameters of an existing index are changed.
        //then drop the index and create a new one
        try {
            dropIndex(keys)
        } catch (e2: Exception) {
            //ignore
        }
        createIndex(keys, indexOptions)
    }
}
```

my questions:
1. `ensureIndex` has been deprecated since [mongo-server-version 3.0.0.](https://docs.mongodb.com/v3.0/reference/method/db.collection.ensureIndex/#db.collection.ensureIndex). Kmongo is using java-driver 4.0, why should it expose this behaviour if it's no longer supported?

2. The [spec definition](https://docs.mongodb.com/v2.6/reference/method/db.collection.ensureIndex/#behaviors) exists for version 2.6, and it clearly specifies that `ensureIndex will not rebuild the existing index with the new options`.  Why is kmongo dropping the old index? I believe alerting a user on an error is a better behaviour here. I would hate to role a version to production and get some of my indices rebuilt without me intending to do so.

**Comment by zigzago** [comment]
You are right, the KMongo `ensureIndex` extension has not the same goal than old mongo method `ensureIndex`.

So it does not follow the mongo spec. In fact, it was designed without knowledge of this mongo method.

The behaviour is [documented](https://litote.org/kmongo/dokka/kmongo/org.litote.kmongo/com.mongodb.client.-mongo-collection/ensure-index.html). It avoids index creation runtime error, when you change the TTL of an index for example. Of course, it is not suitable for all use cases.

Should it be renamed? May be. If you suggest a better name, I would deprecate `ensureIndex`.

Also, I think that a `createIndex` extension with `vararg KProperty<*>` parameter, and a `createUniqueIndex` extension should be added.

**Comment by netag** [comment]
if we renamed it to `createOrRebuild` ? or something along the lines, giving more info to the user besides documentation?
another option is to let the user provide an optional "callback" on how to deal with exceptions happening during the `create` process.
i am more than willing to have a go on such pr, including the extensions you mentioned.

**Comment by zigzago** [comment]
`createOrRebuild` looks good to me. A PR would be welcome :)
